File: hashes/0242-valid-anagram.py
from typing import Counter
import unittest

# https://leetcode.com/problems/valid-anagram/

# python3 -m unittest hashes/0242-valid-anagram.py

class Solution(unittest.TestCase):
    def isAnagram(self, s: str, t: str) -> bool:
        # return sorted(s) == sorted(t)
        return Counter(s) == Counter(t)

    def testIsAnagram(self) -> None:
        for s, t, expected in [
            ("anagram", "nagaram", True),
            ("rat", "car", False),
            ("rrat", "tar", False),
            ("a", "abb", False), # a clear example why bit manipulation doesn't work
        ]:
            output = self.isAnagram(s, t)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: hashes/0242-valid-anagram.go
package hashes

// https://leetcode.com/problems/valid-anagram/

// Approach: Counter
// Time: O(S+T), S=len(s), T=len(t)
// Space: O(26) = O(1)
func isAnagram(s string, t string) bool {
	counter := make(map[rune]int)
	for _, c := range s {
		counter[c]++
	}

	for _, c := range t {
		counter[c]--
	}

	for _, cnt := range counter {
		if cnt != 0 {
			return false
		}
	}

	return true
}

// // Approach: Sorting
// // Time: O(SlogS+TlogT), S=len(s), T=len(t)
// // Space: O(S+T)
// func isAnagram(s string, t string) bool {
// 	sb := []byte(s)
// 	tb := []byte(t)
// 	sort.Slice(sb, func(i, j int) bool { return sb[i] < sb[j] })
// 	sort.Slice(tb, func(i, j int) bool { return tb[i] < tb[j] })
// 	return string(sb) == string(tb)
// }
